auth0/client.go

```diff
@@ -0,0 +1,227 @@
+package auth0
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/mdblp/shoreline/schema"
+	"github.com/pkg/errors"
+)
+
+type userMetaData struct {
+	Role string `json:"role"`
+}
+
+type auth0User struct {
+	Email         string       `json:"email"`
+	EmailVerified bool         `json:"email_verified"`
+	UserId        string       `json:"user_id"`
+	Metadata      userMetaData `json:"user_metadata"`
+}
+
+type auth0Token struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
+}
+
+type Auth0Client struct {
+	// Client id, secret and url are required to establish a connection to Auth0 and get tokens
+	clientId        string
+	secret          string
+	logger          *log.Logger
+	token           string
+	mut             sync.Mutex
+	closed          chan chan bool // Channel to communicate that the object has been closed
+	acquiringToken  bool           // flag set when the serverLoginLoop is running
+	refreshInterval time.Duration  // token refresh period in nanoseconds
+	userUrl         string
+	tokenUrl        string
+	audience        string
+}
+
+func NewAuth0Client(logger *log.Logger) *Auth0Client {
+	auth0Url := os.Getenv("AUTH0_URL")
+	auth0Secret := os.Getenv("AUTH0_Secret")
+	auth0ClientId := os.Getenv("AUTH0_CLIENT_ID")
+	if auth0Url == "" || auth0Secret == "" {
+		logger.Fatal("Auth0 configuration not provided")
+	}
+	interval, _ := time.ParseDuration("1h")
+	usrUrl, err := url.Parse(auth0Url + "/api/v2/users")
+	if err != nil {
+		logger.Fatal("Auth0 configuration incorrect, malformed URL: ", err.Error())
+	}
+	tokenUrl, _ := url.Parse(auth0Url + "/oauth/token")
+	audience, _ := url.Parse(auth0Url + "/api/v2/")
+
+	return &Auth0Client{
+		secret:          auth0Secret,
+		clientId:        auth0ClientId,
+		logger:          logger,
+		refreshInterval: interval,
+		userUrl:         usrUrl.String(),
+		tokenUrl:        tokenUrl.String(),
+		audience:        audience.String(),
+	}
+}
+
+func (client *Auth0Client) Start() error {
+	var err error
+	if client.secret == "" || client.clientId == "" {
+		panic("auth0Client requires a secret and ID to be set")
+	}
+	if err = client.serverLogin(); err != nil {
+		log.Printf("Error on initial server token acquisition, [%v]", err)
+		go client.serverLoginLoop(true)
+	} else {
+		go client.refreshTokenLoop()
+	}
+	return nil
+}
+
+func (client *Auth0Client) serverLoginLoop(launchRefreshTokenLoop bool) {
+	var attempts int64
+	client.mut.Lock()
+	if client.acquiringToken {
+		client.mut.Unlock()
+		return
+	}
+	client.acquiringToken = true
+	client.mut.Unlock()
+	for {
+		timer := time.After(client.refreshInterval)
```

Comment:
We have just one interval for the initial token acquisition and the refresh token ?